### DISCUSSION TITLE
Megadesk Companion Announcement & Discussion

### DISCUSSION THREAD

**Discussion by gcormier** [gcormier/megadesk#136] [General]
[Megadesk Companion](https://github.com/gcormier/megadesk_companion) has released.

It allows for easy `esphome` integration, uses the existing power, and has a 3V/5V logic level shifter to go between the ESP32 and the AVR microcontrollers.

The esphome YAML file is at https://github.com/gcormier/megadesk/blob/master/esphome/megadesk-companion.yaml

Swapping RX/TX is done by swapping these two GPIO's
```
  tx_pin: GPIO6
  rx_pin: GPIO7
```

**Comment by landonr** [comment]
this week i've been working on a completely different companion controller

![image](https://github.com/gcormier/megadesk/assets/2607659/5473fabf-895d-49b4-9eb5-7fd30f12957f)
- it uses a lilygo T-Display to talk to my megadesk, it runs https://homething.io to control music and stuff
- i copied the design of the ikea IDÅSEN controller because I broke my buttons disassembling it
- I'm planning on adding some extra buttons for home assistant shortcuts and a knob for media volume

edit* i need to add a lock because my cat likes to bite it lol

**Reply by landonr** [reply]
here's a video of it in action! https://www.youtube.com/shorts/zrdgeWlkCA0

**Comment by gcormier** [comment]
Here is the default entity in home assistant

![image](https://github.com/gcormier/megadesk/assets/8388983/94f13e13-8af9-4203-be8f-dce353d828a5)

**Reply by landonr** [reply]
I just noticed Desk Position 1 is missing from the top of controls(?)

**Reply by gcormier** [reply]
There is no memory position one as it's not possible to trigger that from the buttons.

**Reply by androidand** [reply]
The reason there is no position 1 is because the way one manually moves the desk vs how a position is stored.
Moving the desk upwards is holding the up button. This would be the same as saving position 1, which is impossible.

**Comment by landonr** [comment]
is it possible to add a software lock to the buttons? So the dashboard could have a switch to lock the desk from moving

**Reply by gcormier** [reply]
Hmm, it should be easy enough. I could add a new command to lock the buttons using a certain # of button pushes. I'll open a issue for that.

**Reply by gcormier** [reply]
Can I assume that the lock would only apply to the physical buttons - serial commands will continue to work regardless of the lock status?

**Reply by landonr** [reply]
I think that would be simplest!

**Comment by landonr** [comment]
It would be nice if the megadesk companion had some extra GPIO for a SPI screen and some buttons to control a ui

**Reply by gcormier** [reply]
How many GPIO do you need?

**Reply by landonr** [reply]
2 for an encoder (scroll wheel), 2 or 3 for buttons, SPI or i2c pins for a display? It kind of depends what people want to use it for

**Reply by gcormier** [reply]
Right now there's 3 GPIO exposed that you could use. I could try to add more solder pads on the underside of the board in the next revision, but there's no room for more headers really.

**Reply by gcormier** [reply]
Oh actually, on V3 of the board, there's 5 GPIO exposed right now which is the max. 1 on LED, 2 via the programming header, and then 2 pads on top expose GPIO4 and GPIO3 as well.

**Reply by landonr** [reply]
that would be sweet! Even a couple pins would be handy! 5 is enough for a fastled rgb strip ;)

**Reply by gcormier** [reply]
That's what is available now - V4 will only be slightly smaller (3mm give or take) to provide easier access to the screws.

**Reply by gcormier** [reply]
V4 has a few pads on top

![image](https://github.com/gcormier/megadesk/assets/8388983/1dad8fb7-6331-44b9-8368-a40773ece6c2)

And two on the bottom
![image](https://github.com/gcormier/megadesk/assets/8388983/5b734780-c0fc-4b23-9be3-b0c4c7431c2e)

It ditches the programming header though, can only be programmed via USB.

**Comment by DJPoulter** [comment]
Does this require a minimum version for the Megadesk? I have the companion fully setup but the desk does not report back a height and none of the buttons do anything 
<img width="1004" alt="Screenshot 2023-09-16 at 16 14 37" src="https://github.com/gcormier/megadesk/assets/24723608/812a881a-2afd-471a-b97a-6aefc001d217">

**Reply by gcormier** [reply]
It requires the SERIAL firmware be flashed - all Megadesks from the store after December 2022 had this. If you bought a megadesk prior to that, it would not have the serial firmware flashed.

https://github.com/gcormier/megadesk/blob/master/DIY.md#programming
https://github.com/gcormier/megadesk/discussions/101

**Reply by DJPoulter** [reply]
I have the megadesk v7. And have tried to swap the tx and rx pins

**Reply by DJPoulter** [reply]
Okay let me get to flashing hopefully 🤞

**Reply by DJPoulter** [reply]
My megadesk does not seem to be starting after flashing. Is there any easy debug steps to work out what has happened? avrdude says flash was perfect. I flashed the v2022.09-t1624-serial.hex

**Reply by gcormier** [reply]
Do you have the ATTiny1624 version of the hardware?

**Reply by unn** [reply]
My megadesk is from April '22 so I'll need to flash the serial firmware. Any chance you could record a short video of that operation or have links to some high quality resources? It'd be pretty helpful, since I'm starting my hardware/mcu hobby. Thanks!
